### PR TITLE
HPCC-11927 Further optimizations to lookup/smart join

### PR DESCRIPTION
**Pull request by jakesmith**
Free stable sort ptr array if in before setting up hash table.
Compact row ptr table after failing over to local lookup.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: thorlcr/thorutil/thmem.cpp [hunk 1]
```diff
@@ -451,7 +428,7 @@ const void *CThorExpandingRowArray::allocateRowTable(rowidx_t num, unsigned maxS
 
 rowidx_t CThorExpandingRowArray::getNewSize(rowidx_t requiredRows)
 {
-    rowidx_t newSize = maxRows;
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: Was this change deliberate?  If so the rest of the function can disappear

File: thorlcr/thorutil/thmem.cpp [hunk 2]
```diff
@@ -543,39 +520,60 @@ inline const void *CThorExpandingRowArray::_allocateRowTable(rowidx_t num, unsig
...
-        CResizeRowCallback callback(*(void ***)(&rows), capacity, queryLock());
-        if (!resizeRowTable((void **)rows, capacity, true, callback, maxSpillCost)) // callback will reset capacity
+        if (numRows > requiredRows)
         {
-            if (throwOnOom)
-                throw MakeActivityException(&activity, 0, "Out of memory, allocating row array, had %"RIPF"d, trying to allocate %"RIPF"d elements", ordinality(), requiredRows);
-            return false;
+            for (rowidx_t i = requiredRows; i < numRows; i++)
+                ReleaseThorRow(rows[i]);
+            numRows -= (numRows-requiredRows);
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: Simpler to say "numRows = requiredRows"!

File: thorlcr/thorutil/thmem.cpp [hunk 3]
```diff
@@ -543,39 +520,60 @@ inline const void *CThorExpandingRowArray::_allocateRowTable(rowidx_t num, unsig
     }
 }
 
-inline bool CThorExpandingRowArray::_ensure(rowidx_t requiredRows, unsigned maxSpillCost)
+inline bool CThorExpandingRowArray::doResize(void **&_rows, rowidx_t requiredRows, bool copy, unsigned maxSpillCost, memsize_t &capacity, const char *errMsg)
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: minor: Might be worth coming up with clearer function name.  resize, _resize and doResize aren't obvious what they do.

File: thorlcr/thorutil/thmem.cpp [hunk 4]
```diff
@@ -588,9 +586,22 @@ inline bool CThorExpandingRowArray::_ensure(rowidx_t requiredRows, unsigned maxS
...
+    stableTable = NULL;
+    rows = NULL;
+    maxRows = 0;
+    numRows = 0;
+    rowManager = activity.queryJob().queryRowManager();
+    throwOnOom = false;
     setup(_rowIf, _allowNulls, _stableSort, _throwOnOom);
     setDefaultMaxSpillCost(roxiemem::SpillAllCost);
-    init(initialSize);
+    if (initialSize)
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: Should this call resize()?

File: thorlcr/thorutil/thmem.cpp [hunk 5]
```diff
@@ -646,7 +662,17 @@ void CThorExpandingRowArray::compact()
         }
     }
     numRows = freeFinger-rows;
-    // JCSMORE - this may be a good time to call IRowManager::compactRows
+    memsize_t numEmptiedPages = rowManager->compactRows(numRows, rows);
+
+#ifdef _DEBUG
+    ActPrintLog(&activity, "CThorExpandingRowArray::compact(): compactRows freed %"I64F"d pages", (unsigned __int64)numEmptiedPages);
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: new code should have spaces between the " and the I64F (C++11 compatibility).  There is a jira open to fix the other instances.

File: thorlcr/thorutil/thmem.cpp [hunk 6]
```diff
@@ -646,7 +662,17 @@ void CThorExpandingRowArray::compact()
...
+#endif
+
+    if (maxSpillCost)
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: Is this test correct?  If this always shrinks, then it should be possible to resize the row array without causing any spills.  Should the parameter be named differently?  Should 0 always be passed to resize?

File: thorlcr/thorutil/thmem.cpp [hunk 7]
```diff
@@ -543,39 +520,60 @@ inline const void *CThorExpandingRowArray::_allocateRowTable(rowidx_t num, unsig
...
+        if (numRows > requiredRows)
         {
-            if (throwOnOom)
-                throw MakeActivityException(&activity, 0, "Out of memory, allocating row array, had %"RIPF"d, trying to allocate %"RIPF"d elements", ordinality(), requiredRows);
-            return false;
+            for (rowidx_t i = requiredRows; i < numRows; i++)
+                ReleaseThorRow(rows[i]);
+            numRows -= (numRows-requiredRows);
         }
+        doResize((void **&)rows, requiredRows, true, maxSpillCost, capacity, "row array");
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: You're not checking the return value from doResize, here and in the following branches.  (Returning false if it fails)

**Comment by ghalliday** [comment]
@jakesmith a few comments/questions

File: thorlcr/thorutil/thmem.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: Well.. it was, but wrong.
It was so it had any effect when reducing in size (ignoring existing maxSize). 
However, I shouldn't have used this function which is for expanding from maxRows size only.

Will fix.

File: thorlcr/thorutil/thmem.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: yep.

File: thorlcr/thorutil/thmem.cpp (hunk 3 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: I'm going to merge doResize into resizeRowTable, that will help.

File: thorlcr/thorutil/thmem.cpp (hunk 4 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: I don't think so.. it _could_ for common code reasons, but more efficient/simpler code to avoid checks/callbacks here in constructor where there is no contention/locking required.

File: thorlcr/thorutil/thmem.cpp (hunk 5 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: Right, I'll change others in this file while I'm at it (one other which is new also)

File: thorlcr/thorutil/thmem.cpp (hunk 6 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: Looking at roxiemem, resizeRow() , I was under the impression that shrinking could allocate in order to shrink... (?)

So I pass in maxSpillCost to control when/what priority I want it to allow spilling at.
Passing 0 indicates that I don't want it to spill under any circumstances to allocate.

File: thorlcr/thorutil/thmem.cpp (hunk 7 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: will fix.

**Comment by jakesmith** [comment]
@ghalliday - one real question back to you, re. spill lcost, shrinking, and roxiemem::resizeRow()'s ability to allocate ...

File: thorlcr/thorutil/thmem.cpp (hunk 6 above)
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: I think the row shrinking was only merged to master (https://github.com/hpcc-systems/HPCC-Platform/pull/6711).  That only shrinks huge pages - which will be done inline.

**Comment by ghalliday** [comment]
@jakesmith see answer

File: thorlcr/thorutil/thmem.cpp (hunk 6 above)
Comment (jakesmith) [inline thorlcr/thorutil/thmem.cpp]: Ok, yes I see that now. I will remove maxSpillCost from compact() etc.

**Comment by jakesmith** [comment]
@ghalliday - please see changes.

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp [hunk 8]
```diff
@@ -1704,7 +1716,8 @@ class CLookupJoinActivityBase : public CInMemJoinBase<HTHELPER, IHThorHashJoinAr
                         CThorSpillableRowArray &rows = *rhsNodeRows.item(a);
                         clearNonLocalRows(rows, flushedRowMarkers.item(a));
 
-                        rows.compact(); // JCS->GH - really we want to resize rhsNodeRows now to free up as much space as possible (see HPCC-12511)
+                        ActPrintLog("Compacting rhsNodeRows[%d], has %"RIPF"d rows", a, rows.numCommitted());
```
Comment (ghalliday) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: minor: spaces needed around RIPF

File: thorlcr/thorutil/thmem.cpp [hunk 9]
```diff
@@ -468,25 +445,43 @@ rowidx_t CThorExpandingRowArray::getNewSize(rowidx_t requiredRows)
...
-    catch (IException * e)
+    catch (IException *e)
     {
         //Pathological cases - not enough memory to reallocate the target row buffer, or no contiguous pages available.
         unsigned code = e->errorCode();
         if ((code == ROXIEMM_MEMORY_LIMIT_EXCEEDED) || (code == ROXIEMM_MEMORY_POOL_EXHAUSTED))
         {
             e->Release();
+            if (throwOnOom)
+                throw MakeActivityException(&activity, 0, "Out of memory, resizing %s, had %"RIPF"d, trying to allocate %"RIPF"d elements", errMsg, ordinality(), requiredRows);
```
Comment (ghalliday) [inline thorlcr/thorutil/thmem.cpp]: minor spacing again

**Comment by ghalliday** [comment]
@jakesmith forgot to reassign for minor changes.

**Comment by jakesmith** [comment]
@ghalliday - added missing spaces in these 2 files. I'll open another PR to correct the reset in other files in Thor.

**Comment by jakesmith** [comment]
@ghalliday - rebased.
